src/core/src/bound_evaluate.cpp

```diff
@@ -256,9 +280,59 @@ ov::Tensor ov::evaluate_upper_bound(const Output<Node>& output) {
 }
 
 std::pair<ov::Tensor, ov::Tensor> ov::evaluate_both_bounds(const Output<Node>& output) {
-    evaluate_bound(output, false, false);
-    evaluate_bound(output, true);
-    return {output.get_tensor_ptr()->get_lower_value(), output.get_tensor_ptr()->get_upper_value()};
+    const auto& output_tensor = output.get_tensor();
+    if (output_tensor.get_lower_value() && output_tensor.get_upper_value())
+        return {output_tensor.get_lower_value(), output_tensor.get_upper_value()};
+
+    std::vector<Node*> order;
+    if (could_propagate(output, order)) {
+        for (const auto& node : order) {
+            ov::TensorVector outputs_lower, outputs_upper;
+            for (const auto& out : node->outputs()) {
+                OPENVINO_SUPPRESS_DEPRECATED_START
+                outputs_lower.push_back(util::wrap_tensor(out));
+                outputs_upper.push_back(util::wrap_tensor(out));
+                OPENVINO_SUPPRESS_DEPRECATED_END
+            }
+
+            if (!node->evaluate_lower(outputs_lower) || !node->evaluate_upper(outputs_upper)) {
+                break;
+            }
+            auto input_values = node->input_values();
+            bool same_inputs = std::all_of(input_values.begin(), input_values.end(), [](const Output<Node>& input) {
+                auto& t = input.get_tensor();
+                return t.has_and_set_bound() || are_equal(t.get_lower_value(), t.get_upper_value());
+            });
+
+            for (size_t i = 0; i < node->get_output_size(); ++i) {
+                auto& out_tensor = node->get_output_tensor(i);
+                out_tensor.set_lower_value(outputs_lower[i]);
+                out_tensor.set_upper_value(outputs_upper[i]);
+                if (same_inputs || are_equal(outputs_lower[i], outputs_upper[i])) {
+                    out_tensor.set_upper_value(outputs_lower[i]);
+                }
+            }
+            TensorLabelVector output_labels(node->get_output_size());
+            if (node->evaluate_label(output_labels))
+                for (size_t i = 0; i < output_labels.size(); ++i)
+                    node->get_output_tensor(i).set_value_label(output_labels[i]);
+
+            for (const auto& input : node->input_values()) {
+                auto& tensor = input.get_tensor();
+                bool should_invalidate = false;
+                const auto& lower = tensor.get_lower_value();
+                const auto& upper = tensor.get_upper_value();
+                if (lower && shape_size(lower.get_shape()) > 10)
+                    should_invalidate |= true;
+                if (upper && shape_size(upper.get_shape()) > 10)
+                    should_invalidate |= true;
```

Comment:
```suggestion
                const auto should_invalidate = (lower && shape_size(lower.get_shape()) > 10) || (upper && shape_size(upper.get_shape());
```
Just detail to avoid bitwise operation on bools, and flag will be set once instead 3 times in worst case.